File: test/shire/projects_test.exs
defmodule Shire.ProjectsTest do
  use Shire.DataCase, async: true

  alias Shire.Projects

  describe "create_project/1" do
    test "creates a project with valid name" do
      assert {:ok, project} = Projects.create_project("my-project")
      assert project.name == "my-project"
      assert project.id
    end

    test "rejects duplicate names" do
      {:ok, _} = Projects.create_project("unique-name")
      assert {:error, changeset} = Projects.create_project("unique-name")
      assert "has already been taken" in errors_on(changeset).name
    end

    test "requires name" do
      assert {:error, changeset} = Projects.create_project(nil)
      assert "can't be blank" in errors_on(changeset).name
    end
  end

  describe "get_project!/1" do
    test "returns project by id" do
      {:ok, project} = Projects.create_project("get-test")
      assert Projects.get_project!(project.id).name == "get-test"
    end

    test "raises on missing id" do
      assert_raise Ecto.NoResultsError, fn ->
        Projects.get_project!(Ecto.UUID.generate())
      end
    end
  end

  describe "get_project_by_id/1" do
    test "returns project or nil" do
      {:ok, project} = Projects.create_project("by-id-test")
      assert Projects.get_project_by_id(project.id).name == "by-id-test"
      assert Projects.get_project_by_id(Ecto.UUID.generate()) == nil
    end
  end

  describe "get_project_by_name/1" do
    test "returns project or nil" do
      {:ok, _} = Projects.create_project("by-name-test")
      assert Projects.get_project_by_name("by-name-test").name == "by-name-test"
      assert Projects.get_project_by_name("nonexistent") == nil
    end
  end

  describe "list_projects/0" do
    test "returns projects ordered by name" do
      {:ok, _} = Projects.create_project("zeta")
      {:ok, _} = Projects.create_project("alpha")
      {:ok, _} = Projects.create_project("mid")

      names = Projects.list_projects() |> Enum.map(& &1.name)
      assert names == ["alpha", "mid", "zeta"]
    end
  end

  describe "rename_project/2" do
    test "updates the project name" do
      {:ok, project} = Projects.create_project("old-name")
      assert {:ok, renamed} = Projects.rename_project(project, "new-name")
      assert renamed.name == "new-name"
      assert Projects.get_project!(project.id).name == "new-name"
    end
  end

  describe "delete_project/1" do
    test "deletes the project" do
      {:ok, project} = Projects.create_project("delete-me")
      assert {:ok, _} = Projects.delete_project(project)
      assert Projects.get_project_by_id(project.id) == nil
    end
  end
end

File: src/services/projects.test.ts
import { describe, it, expect } from "bun:test";
import { useTestDb } from "../test/setup";
import * as projects from "./projects";

describe("projects service", () => {
  useTestDb();

  describe("createProject", () => {
    it("creates a project with valid name", () => {
      const project = projects.createProject("my-project");
      expect(project.name).toBe("my-project");
      expect(project.id).toBeTruthy();
    });

    it("rejects duplicate names", () => {
      projects.createProject("unique-name");
      expect(() => projects.createProject("unique-name")).toThrow();
    });
  });

  describe("getProject", () => {
    it("returns project by id", () => {
      const project = projects.createProject("get-test");
      const found = projects.getProject(project.id);
      expect(found?.name).toBe("get-test");
    });

    it("returns undefined for missing id", () => {
      expect(projects.getProject("nonexistent-id")).toBeUndefined();
    });
  });

  describe("getProjectByName", () => {
    it("returns project by name", () => {
      projects.createProject("by-name-test");
      expect(projects.getProjectByName("by-name-test")?.name).toBe("by-name-test");
    });

    it("returns undefined for missing name", () => {
      expect(projects.getProjectByName("nonexistent")).toBeUndefined();
    });
  });

  describe("listProjects", () => {
    it("returns projects ordered by name", () => {
      projects.createProject("zeta");
      projects.createProject("alpha");
      projects.createProject("mid");

      const names = projects.listProjects().map((p) => p.name);
      expect(names).toEqual(["alpha", "mid", "zeta"]);
    });
  });

  describe("renameProject", () => {
    it("updates the project name", () => {
      const project = projects.createProject("old-name");
      const renamed = projects.renameProject(project.id, "new-name");
      expect(renamed?.name).toBe("new-name");
      expect(projects.getProject(project.id)?.name).toBe("new-name");
    });
  });

  describe("deleteProject", () => {
    it("deletes the project", () => {
      const project = projects.createProject("delete-me");
      projects.deleteProject(project.id);
      expect(projects.getProject(project.id)).toBeUndefined();
    });
  });
});
